lgc/patch/PatchBufferOp.cpp

```diff
@@ -64,129 +79,284 @@ PreservedAnalyses PatchBufferOp::run(Function &function, FunctionAnalysisManager
+  DenseSet<Value *> seen;
+  SmallVector<Value *> searchWorklist;
+  SmallVector<Value *> propagationWorklist;
+  Value *current = desc;
+  for (;;) {
+    if (seen.insert(current).second) {
+      auto &di = m_descriptors[current];
+
+      if (!di.invariant.has_value() || !di.divergent.has_value()) {
+        if (auto *phi = dyn_cast<PHINode>(current)) {
+          for (Value *incoming : phi->incoming_values())
+            searchWorklist.push_back(incoming);
+        } else if (auto *select = dyn_cast<SelectInst>(current)) {
+          assert(select->getOperandUse(0).get() == select->getCondition());
+          if (m_divergenceInfo.isDivergentUse(select->getOperandUse(0)))
+            di.divergent = true;
+
+          if (!di.invariant.has_value() || !di.divergent.has_value()) {
+            searchWorklist.push_back(select->getTrueValue());
+            searchWorklist.push_back(select->getFalseValue());
+          }
+        } else {
+          // Make conservative assumptions for unhandled instructions.
+          bool isConstant = isa<Constant>(current);
+          auto &di = m_descriptors[current];
+          if (!di.invariant.has_value())
+            di.invariant = isConstant;
+          if (!di.divergent.has_value()) {
+            // TODO: This would be entirely unnecessary if we had updatable divergence info.
+            di.divergent = !isConstant;
+          }
+        }
+      }
+
+      if (!di.invariant.value_or(true) || di.divergent.value_or(false))
+        propagationWorklist.push_back(current);
+    }
+
+    if (searchWorklist.empty())
+      break;
+
+    current = searchWorklist.pop_back_val();
   }
-  m_postVisitInsts.clear();
 
-  const bool changed = (!m_replacementMap.empty());
+  // Fixed-point iteration to propagate "variant" and "divergent" flags.
+  while (!propagationWorklist.empty()) {
+    current = propagationWorklist.pop_back_val();
+    auto diIt = m_descriptors.find(current);
+    assert(diIt != m_descriptors.end());
+    DescriptorInfo di = diIt->second;
+
+    for (User *user : current->users()) {
+      // Make a reasonable effort not to "leak" into instructions we don't understand (e.g., if a pointer / descriptor
+      // ended up in an aggregate). Some of these cases could perhaps be handled in a conservative way, but it seems
+      // unlikely to be necessary in practice.
+      if (!isa<PHINode>(user) && !isa<SelectInst>(user) && !seen.count(user))
+        continue;
 
-  for (auto &replaceMap : m_replacementMap) {
-    Instruction *const inst = replaceMap.first;
-    inst->replaceAllUsesWith(UndefValue::get(inst->getType()));
-    inst->eraseFromParent();
+      auto &userDi = m_descriptors[user];
+      bool propagate = false;
+      if (!userDi.invariant.has_value() && di.invariant.has_value()) {
+        userDi.invariant = di.invariant;
+        if (!di.invariant.value())
+          propagate = true;
+      }
+      if (!userDi.divergent.has_value() && di.divergent.has_value()) {
```

Comment:
I find another corner case might have issue, but I think we are pretty much close to the final version:) Think a select with uniform condition and one input descriptor is uniform, the other descriptor is divergent. If the uniform source descriptor firstly propagates its `divergent:false` to the select because it was propagating variant property. Then the divergent descriptor will not be able to propagate the `divergent:true` to the select, right? The variant property propagation would also be affected in the opposite way.